sui_core/src/gateway_types.rs

```diff
@@ -0,0 +1,947 @@
+        match value {
+            ObjectRead::NotExists(id) => Ok(SuiObjectRead::NotExists(id)),
+            ObjectRead::Exists(_, o, layout) => {
+                Ok(SuiObjectRead::Exists(SuiObject::try_from(o, layout)?))
+            }
+            ObjectRead::Deleted(oref) => Ok(SuiObjectRead::Deleted(oref)),
+        }
+    }
+}
+
+#[serde_as]
+#[derive(Debug, Deserialize, Serialize, JsonSchema, Clone, Eq, PartialEq)]
+#[serde(untagged, rename = "MoveValue")]
+pub enum SuiMoveValue {
+    // Move base types
+    U8(u8),
+    U64(u64),
+    U128(u128),
+    Bool(bool),
+    Address(ObjectID),
+    Vector(Vec<SuiMoveValue>),
+    Struct(SuiMoveStruct),
+    Signer(SuiAddress),
+
+    // Sui base types
+    String(String),
+    ID(ID),
+    UniqueID(UniqueID),
+    VersionedID(VersionedID),
+    Balance(u64),
+    ByteArray(Base64),
+    Coin(Coin),
+}
+
+impl From<MoveValue> for SuiMoveValue {
+    fn from(value: MoveValue) -> Self {
+        match value {
+            MoveValue::U8(value) => SuiMoveValue::U8(value),
+            MoveValue::U64(value) => SuiMoveValue::U64(value),
+            MoveValue::U128(value) => SuiMoveValue::U128(value),
+            MoveValue::Bool(value) => SuiMoveValue::Bool(value),
+            MoveValue::Address(value) => SuiMoveValue::Address(ObjectID::from(value)),
+            MoveValue::Vector(value) => {
+                if value.iter().all(|value| matches!(value, MoveValue::U8(_))) {
+                    let bytearray = value
+                        .iter()
+                        .flat_map(|value| {
+                            if let MoveValue::U8(u8) = value {
+                                Some(*u8)
+                            } else {
+                                None
+                            }
+                        })
+                        .collect::<Vec<_>>();
+                    return SuiMoveValue::ByteArray(Base64::from_bytes(&bytearray));
+                }
+                SuiMoveValue::Vector(value.into_iter().map(|value| value.into()).collect())
+            }
+            MoveValue::Struct(value) => {
+                // Best effort Sui core type conversion
+                if let MoveStruct::WithTypes { type_, fields } = &value {
+                    if let Some(value) = try_convert_type(type_, fields) {
+                        return value;
+                    }
+                };
+                SuiMoveValue::Struct(value.into())
+            }
+            MoveValue::Signer(value) => {
+                SuiMoveValue::Signer(SuiAddress::from(ObjectID::from(value)))
+            }
+        }
+    }
+}
+
+#[derive(Debug, Deserialize, Serialize, JsonSchema, Clone, Eq, PartialEq)]
+#[serde(untagged, rename = "MoveStruct")]
+pub enum SuiMoveStruct {
+    Runtime(Vec<SuiMoveValue>),
+    WithFields(Vec<(String, SuiMoveValue)>),
+    WithTypes {
+        #[serde(rename = "type")]
+        type_: String,
+        fields: BTreeMap<String, SuiMoveValue>,
+    },
+}
```

Comment:
@patrickkuo , I am a bit confused about why we need three enum values here. Could you clarify in which cases does the client side will receive each of these? 

I found the deserialized schema are not very consistent:

For example, this one has `type` and `fields` in `contents`
```
"details": {
      "data": {
        "dataType": "moveObject",
        "type": "0x2::DevNetNFT::DevNetNFT",
        "contents": {
          "type": "0x2::DevNetNFT::DevNetNFT",
          "fields": {
            "description": "An NFT created by the wallet Command Line Tool",
            "id": {
              "id": "0xaadd82329e47e73849efe6ffe7296de675adff03",
              "version": 1
            },
            "name": "Example NFT",
            "url": "ipfs://bafkreibngqhl3gaa7daob4i2vccziay2jjlp435cf66vhono7nrvww53ty"
          }
        }
      }
```

But this one does not have `type` and `fields` in `contents`

```
"details": {
      "data": {
        "dataType": "moveObject",
        "type": "0x2::Coin::Coin<0x2::SUI::SUI>",
        "contents": {
          "id": {
            "id": "0x2213b35d051547f1c559e5988507788e2e6710f7",
            "version": 1
          },
          "value": 99241
        }
      }
```

The inconsistency makes the client side parsing very hard. Can we either get rid of `type` and `fields` in `contents`(preferred, since we already have a `type` field) or always have them?

Reply:
This is not nice ......
I will lift `MoveStruct::with_type` out if it's the top level struct (MoveObject's data structure is = MoveStruct::with_type)
This should make the response consistent across MoveValue/MoveStructs

Reply:
the response looks like this after the changes

```
{
        "jsonrpc": "2.0",
        "result": {
            "status": "Exists",
            "details": {
                "data": {
                    "dataType": "moveObject",
                    "type": "0x2::Coin::Coin<0x2::SUI::SUI>",
                    "contents": {
                        "balance": 98125,
                        "id": {
                            "id": "0x2778e353bd9bc2808242a3e3f17d300b0b65a25f",
                            "version": 6
                        }
                    }
                },
                "owner": {
                    "AddressOwner": "0x6cd65bff0ee3c5c7d268d1bbd2db1848f8f3b406"
                },
                "previousTransaction": "Ox+KSE8Mhn1MN+GpjlYeWuV1RLdsagDmvLrkq+8tMYw=",
                "storageRebate": 15,
                "reference": {
                    "objectId": "0x2778e353bd9bc2808242a3e3f17d300b0b65a25f",
                    "version": 6,
                    "digest": "Yqp1gokn9J3AoZPid4U1tz/D6iEvNsgsxWXvEnSLeLw="
                }
            }
        },
        "id": 0
    },
    {
        "jsonrpc": "2.0",
        "result": {
            "status": "Exists",
            "details": {
                "data": {
                    "dataType": "moveObject",
                    "type": "0x2::DevNetNFT::DevNetNFT",
                    "contents": {
                        "description": "An NFT created by the wallet Command Line Tool",
                        "id": {
                            "id": "0x0fb1738d2a8f22d8abd5f29fd920c03c4a38c0de",
                            "version": 1
                        },
                        "name": "Example NFT",
                        "url": "ipfs://bafkreibngqhl3gaa7daob4i2vccziay2jjlp435cf66vhono7nrvww53ty"
                    }
                },
                "owner": {
                    "AddressOwner": "0x6cd65bff0ee3c5c7d268d1bbd2db1848f8f3b406"
                },
                "previousTransaction": "mdW/pCiQqMdwL7H39pShBlDwUYvD9HKgfWnxDpCcFKc=",
                "storageRebate": 25,
                "reference": {
                    "objectId": "0x0fb1738d2a8f22d8abd5f29fd920c03c4a38c0de",
                    "version": 1,
                    "digest": "UehpO6QyBYVtLXq+w8ZCkEiZBZcyJVcVzKX62uxAIc8="
                }
            }
        },
        "id": 1
    }
```